### PR TITLE
Add CSI VolumeCreate node round-robin

### PR DESCRIPTION
**Pull request by ggriffiths**
Signed-off-by: Grant Griffiths <ggriffiths@purestorage.com>
**What this PR does / why we need it**:
Add CSI VolumeCreate round-robbin

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by ggriffiths** [comment]
CSI e2e test passed:
https://jenkins.pwx.dev.purestorage.com/job/Torpedo/job/tp-nextpx-op-csi-e2e/185/

**Comment by ggriffiths** [comment]
torpedo running
https://jenkins.pwx.dev.purestorage.com/job/Dev/job/torpedo-dev01/134/

File: csi/csi.go [hunk 1]
```diff
@@ -155,6 +166,48 @@ func (s *OsdCsiServer) getConn() (*grpc.ClientConn, error) {
...
+		targetNodeNumber = s.nextCreateNodeNumber
+	}
+	targetNodeEndpoint := nodesResp.Nodes[targetNodeNumber].MgmtIp
+	s.nextCreateNodeNumber = (targetNodeNumber + 1) % len(nodesResp.Nodes)
+
+	// Get conn for this node, otherwise create new conn
+	if len(s.connMap) == 0 {
+		s.connMap = make(map[string]*grpc.ClientConn)
+	}
+	if s.connMap[targetNodeEndpoint] == nil {
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply csi/csi.go:194]: I did some research and each gRPC-go client conn is a connection pool that will retry and refresh a connection.

This is why we don't need to do any conn refreshing in the existing implementation.

However, I'm adding a grpc-go keepalive dial option to minimize this conn refreshing.

File: csi/controller_test.go [hunk 2]
```diff
@@ -1610,13 +1710,110 @@ func TestControllerCreateVolumeNoNewVolumeInfo(t *testing.T) {
...
+	s.MockCluster().EXPECT().
+		Enumerate().
+		Return(api.Cluster{
+			NodeId: "node-1",
+			Nodes: []*api.Node{{
+				Id:     "1",
+				MgmtIp: "[::]:",
+			}},
+		}, nil).
+		AnyTimes()
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply csi/controller_test.go]: Updated this

File: csi/csi.go [hunk 3]
```diff
@@ -155,6 +166,48 @@ func (s *OsdCsiServer) getConn() (*grpc.ClientConn, error) {
...
+	// Get target node info and set next round robbin node
+	var targetNodeNumber int
+	if s.nextCreateNodeNumber != 0 {
+		targetNodeNumber = s.nextCreateNodeNumber
+	}
+	targetNodeEndpoint := nodesResp.Nodes[targetNodeNumber].MgmtIp
+	s.nextCreateNodeNumber = (targetNodeNumber + 1) % len(nodesResp.Nodes)
+
+	// Get conn for this node, otherwise create new conn
+	if len(s.connMap) == 0 {
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply csi/csi.go:198]: Updated

**Review by pureneelesh — APPROVED** [review]

File: csi/csi.go [hunk 4]
```diff
@@ -125,12 +132,17 @@ func NewOsdCsiServer(config *OsdCsiServerConfig) (grpcserver.Server, error) {
 		return nil, fmt.Errorf("Failed to create CSI server: %v", err)
 	}
 
+	sdkPort := config.SdkPort
+	if len(config.SdkPort) == 0 {
```
Comment (adityadani) [inline csi/csi.go:136]: This seems like a bug if you are seeing a config.SdkPort being set to 0. Was this required just unit tests?

File: csi/csi.go [hunk 5]
```diff
@@ -155,6 +173,66 @@ func (s *OsdCsiServer) getConn() (*grpc.ClientConn, error) {
...
+
+	// Clean up connections for missing node
+	nodesMap := make(map[string]bool)
+	for _, node := range nodesResp.Nodes {
+		nodesMap[node.MgmtIp] = true
+	}
+	for ip := range s.connMap {
+		// If key in connmap is not in current nodes, remove it
+		if ok := nodesMap[ip]; !ok {
+			delete(s.connMap, ip)
```
Comment (adityadani) [inline csi/csi.go:197]: We should close the grpc connection as well? We have seen memory leaks where not closing a connection would result into leaking goroutines.

**Comment by adityadani** [comment]
I am a bit on the fence to enable keepalive for the remote gRPC connections. On a large scale cluster, this could potentially mean we have n*n-1 connections on each node. (We will start with 3 nodes where the px-csi-ext pods run but if these pods get deleted and get schedule on new nodes then those nodes will create gRPC connections but the old nodes will not remove the unused gRPC connections.)
CreateVolume is relatively less invoked API as compared to others like Attach/Mount. I am thinking it would be better if we simply create a new connection on every create call, let K creates go through it and then kill the connection and on the next (k+1)th create call create a gRPC connection to the next node. Wdyt ?

**Comment by ggriffiths** [comment]
Discussed offline, we will have a follow up issue to address connection pooling cleanup

File: csi/csi.go (hunk 4 above)
Comment (ggriffiths) [reply csi/csi.go:136]: This was for CSI Sanity test or unit test IIRC.. I can configure them on the test side though. Will make a note for the followup issue

**Pull request merged**